lib/providers/portion/PortionProvider.ts

```diff
@@ -0,0 +1,255 @@
+      // also the tokenIn and tokenOut can be native token
+      // portion service only accepts wrapped token address
+      return await Promise.all([
+        this.tokenFetcher.resolveTokenBySymbolOrAddress(sharedInfo.tokenInChainId, sharedInfo.tokenIn),
+        this.tokenFetcher.resolveTokenBySymbolOrAddress(sharedInfo.tokenOutChainId, sharedInfo.tokenOut),
+      ]);
+    } catch (e) {
+      log.error({ e }, 'Failed to resolve tokenIn & tokenOut');
+      metrics.putMetric(`PortionProvider.resolveTokenErr`, 1);
+
+      return [undefined, undefined];
+    }
+  }
+
+  private getPortionAmount(
+    tokenOutCurrencyAmount?: CurrencyAmount<Currency>,
+    portionResponse?: GetPortionResponse,
+    resolvedTokenOut?: Currency
+  ): CurrencyAmount<Currency> | undefined {
+    return portionResponse?.hasPortion && portionResponse.portion && resolvedTokenOut && tokenOutCurrencyAmount
+      ? tokenOutCurrencyAmount.multiply(new Fraction(portionResponse.portion.bips, 10000))
+      : undefined;
+  }
+
+  private getPortionAdjustedTokenOutAmount(
+    tradeType: TradeType,
+    tokenOutCurrencyAmount?: CurrencyAmount<Currency>,
+    portionAmount?: CurrencyAmount<Currency>
+  ): CurrencyAmount<Currency> | undefined {
+    if (tokenOutCurrencyAmount && portionAmount) {
+      switch (tradeType) {
+        case TradeType.EXACT_INPUT:
+          return tokenOutCurrencyAmount.subtract(portionAmount);
+        case TradeType.EXACT_OUTPUT:
+          return tokenOutCurrencyAmount.add(portionAmount);
+        default:
+          // unknown trade type.
+          // instead of throw, just return original tokenOutAmount
+          return tokenOutCurrencyAmount;
+      }
+    }
+
+    return tokenOutCurrencyAmount;
+  }
+
+  // despite the convoluted math, this private method is not important since it's not affecting the actual swap,
+  // nor does clients use any data from this method.
+  private getQuoteGasAndPortionAdjusted(
+    tradeType: TradeType,
+    portionAmount?: CurrencyAmount<Currency>,
+    quoteCurrencyAmount?: CurrencyAmount<Currency>,
+    quoteGasAdjustedCurrencyAmount?: CurrencyAmount<Currency>,
+    requestCurrencyAmount?: CurrencyAmount<Currency>
+  ): CurrencyAmount<Currency> | undefined {
+    switch (tradeType) {
+      case TradeType.EXACT_INPUT:
+        return quoteGasAdjustedCurrencyAmount && portionAmount
+          ? quoteGasAdjustedCurrencyAmount.subtract(portionAmount)
+          : quoteGasAdjustedCurrencyAmount;
+      case TradeType.EXACT_OUTPUT:
+        return this.getQuoteGasAndPortionAdjustedForExactOut(
+          portionAmount,
+          quoteCurrencyAmount,
+          quoteGasAdjustedCurrencyAmount,
+          requestCurrencyAmount
+        );
+      default:
+        // unknown trade type.
+        // instead of throw, just return original quoteGasAdjustedCurrencyAmount
+        return quoteGasAdjustedCurrencyAmount;
+    }
+  }
+
+  private getQuoteGasAndPortionAdjustedForExactOut(
```

Comment:
yep so in my previous comment about returning early if something is undefined would allow for the arguments here to not have to be optional, and you don't have to do a conditional definedness check on every single line